readme.md
**Project Status: Inactive**

Fan Update
---------------------------------------
* Disabled Talents, Guild and Feats
* PVP module support for WoW Classic
* Support for ranged slot
* Item level fix
* Sort average item level fix
* Item scan for hit/crit % and weapon skills
* Fixed enchant name in gearlist module
* Character stats should be display correctly even if the client didn't cached the item yet (workaround)
* Currently support English, French, Spanish, Korean, Chinese and Deutsch clients
* Ability to Show/Hide PVP and gearlist module in options
* Added back Item Usage module for fun statistics :)

To do:
* Portuguese and Russian translation
* Fix stats in the cache module, you might have to enter the same entry twice to show all the stats

Known issues:
* HonorSpy break Examiner, see [Issue 6](https://github.com/taxidriveer/Examiner/issues/6#issuecomment-578391763) 

![alt text](https://i.imgur.com/voq7odP.jpg)

Examiner - An Advanced Inspection Addon
---------------------------------------
With Examiner, you can check other players gear, talents, achievements, honor and arena team details.

When you inspect someone, it will show a stat summery of all their equipped gear combined.
Please note that these values are from gear alone, and will not include bonuses from buffs, talents or normal base stats.

Each player you inspect can be cached, that way you can look them up later, even when they are not around.
The option to cache others is disabled by default, you will have to enable caching on the "Config" page.

You can bind a key to inspect your target, you can even bind a key to inspect whoever is under your mouse, without having to target anyone (stealth inspection).
To set this up, open the Key Bindings dialog and look under "Examiner".

Examiner does not use that much memory by itself, but once you start caching a lot of people, memory usage can grow quite quickly.

You can always find the latest release of this addon on Curse or WoWInterface.
Post your comments, questions and suggestions for this addon on the addons comment page.
Or in this thread: http://forums.curseforge.com/showthread.php?t=18999

Comparing Gear
--------------
To compare one person's gear with another, you can mark a target for compare, by right clicking on the "Stats" button to get the drop down menu to open,
this menu will have an entry called "Mark for Compare". You can also use the shortcut by holding down Shift when clicking the "Stats" button.
When marked for compare, the Stats page will no longer show the actual stats of your inspected person, but will instead show the difference in stats
compared to the person you marked for compare.

Cache Filtering
---------------
If your number of cached entries is getting quite high, you may want a way to filter out specific players.
First you have to open the filter dialog, do this by right clicking on the Cache button and pick "Filter...", or use the shortcut by holding down the Shift key and then click the Cache button.
The filter works a bit similar to the /who function in WoW, so you can for example set the filter to "1-79" and it will show all non level 80s.
To only show a specific class, you can use the filter "class:deathknight", and only Death Knights will be shown.
You can add as many filter matches as you want, so using the filter: "class:hunter race:nightelf guild:awesome" will show all Night Elf hunters from the guild <Awesome>.
Filter matches cannot have spaces, you must write them without, that is why "Death Knight" has to be written as "DeathKnight", you can even just use "deathk" it does not have to match the entire thing.
You can use the filter type "sex" to match males or females. Using "sex:2" will match males, "sex:3" will match females.
There are many filter types, here are some examples: "class", "race", "realm", "guild", "guildRank", "sex" and "zone".

Model Frame Controls
--------------------
Left Click + Move:		Rotates the Model
Right Click + Move:		Moves the Model
Mousewheel:				Zoom
Ctrl + Left Click:		Change Background
Ctrl + Right Click:		Toggle Background
Shift + Left Click:		Reset Model Position 

Slash Commands
--------------
The slash command for Examiner is /examiner or just /ex.

Although you probably wont need to use any slash commands as almost everything is available from the UI,
there are a few things which can only be done through slash commands.

/ex inspect <unit> or /ex i 'unit'
This one will inspect the given unit (target, focus, party3 etc). If no unit token is given, it will inspect target and then player if no target exist.

/ex si <itemLink>
This command will scan just a single item and list its combined stats in the chat frame.
Holding down ALT while the mouse is over an item in Examiner will show a tooltip with the stats like this.

/ex compare <itemLink1> <itemLink2>
Compares two items and lists the stat differences.

/ex arena <rating>
Shows you the amount of arena points the given rating gives for 2v2, 3v3 and 5v5.

/ex rating <stat> <rating> <level>
Converts a given stat rating into the amount it gives in percentage, at the specified level. Will use your level if omitted.

/ex scale <value>
Changes the scale of the Examiner window.

/ex reset
Resets the position in case it was moved off screen.

/ex clearcache
Clears the entire cache of Examiner.

Special Thanks
--------------
- Chester, the original author of SuperInspect, who gave me the idea to make this mod.
- Haldamir of Gorgonnash, for the German translation.
- Qcat0214, who made the Chinese translation for the 1.13.2 patch.
- omosiro, who made the Korean translation.
- chkid, for Korean translation update for the 1.13.2 patch.
- g3gg0, changes to the German translation & author to one of the SuperInspect versions.
- Pettigrow of Sinstralis, for the translation to French.
- Siphony of EU-Onyxia, updated German translation.
- Dridzt, for some of the required changes in the WotLK to make it function properly again.
- quicksylver, for updating the English patterns to WotLK.
- Darigaaz of SoulFlayer(RU) for the Russian translation.
- Hoofik of AshenVale for the 3.0 patch update of the Russian translation.
- Maxfunkey (aka Carambha of EU-Lordearon), patch 3.0.2 update of the German and Frensh translation.
- Thorakon (aka Pulgrim of EU-Alexstrasza), for the patch 3.0.3 update of the German translation.
- ericyen, for the Taiwan translation for the 3.1 release.
- eusi, for the updated talent module in Mists of Pandaria.
- ywfn, for making the PvP module work with all pvp and battleground changes.
- fallendos, for an update to the Korean patterns.
- Netrox, for the help in updaing the talents module for WoD.
- Cidrei, updates to the GetUpgradedItemLevel lib to fix itemLevels for Timewarped items.
- Valmere for ZG enchants and more.
- vanillalad for Details talents implementation and AQ enchant (enUS).
- All the people using Examiner!

Lacking Features, Ideas & Problems
----------------------------------
- Always listen for INSPECT_READY, and invalidate the current unit if an event is recieved that is not "our" unit.
- Cache: Modify the function that builds a list of players from the filter, to take a table parameter, so external code can request a filterd list, without modifying the cache list.
- Postpone the OnCache() event, currently, the OnAchievementsReady() happens after caching is done, so it doesn't actually get cached!
- Module's OnInitialize() shouldn't run until first time Examiner is opened. This would allow for much more optimization.
- mod:HasData() should be posting an event. It messes with the modlist module.
- Locked frames seem to have moved a bit up as of 4.0, change this in Examiner to so it will align with other frames.
- Bring back the ability, as a new module, to send an inspected player to another person with Examiner.


changelog.txt
Examiner Changes | 2020 Dec 24
- TOC update

Examiner Changes | 2019 Nov 23
- Added zhCN localization, thanks to Qcat0214

Examiner Changes | 2019 Nov 21
- PVP module fixes (Progress bar and rank)

Examiner Changes | 2019 Nov 17
------------------------------
- Disabled Talents, Guild and Feats modules
- PVP module support for WoW Classic
- Support for ranged slot
- Item level fix
- Sort average item level fix
- Item scan for hit/crit % and weapon skills
- Fixed enchant name in gearlist module
- Character stats should be display correctly even if the client didn't cached the item yet (workaround)
- Currently support English, French, Spanish, Korean and Deutsch clients
- Ability to Show/Hide PVP and gearlist module in options
- Added back Item Usage module for fun statistics

Examiner Changes | 2018 Aug 12
------------------------------
- Improved the ItemString breakdown when holding down [Alt + Shift] on an item slot.
- Removed the "Examiner.xml" file, all required files are now referenced from the toc file.
- Stats: Removed "Reforged Items" from the details page.
- DelayedScan: Fixed "Delayed Gear Scan Fix" not being able to be turned off.
- Updated the AzDropDown and LibItemString libraries.

Examiner Changes | 2018 Jul 25
------------------------------
- Updated the TOC version to 80000 for the Battle for Azeroth 8.0 pre-patch.
- Fixed the PlaySound() API call, which was changed in the 7.3.0 patch.
- Feats: Disabled wordwrap for the achievement names and descriptions to fix wrapping issues.

Examiner Changes | 2016 Aug 30
------------------------------
- GetUpgradedItemLevel: This library has been renamed into LibItemString. It has a few more features as well as using tooltip scanning to get the true itemLevel.

Examiner Changes | 2016 Aug 24
------------------------------
- GetUpgradedItemLevel: Fixed upgradeID pattern not having a capture group.

Examiner Changes | 2016 Aug 22
------------------------------
- TOC version updated to 70000 for the Legion 7.0.3 pre-patch.
- Talents: Updated this module so it should be working again (thanks Farami).
- Talents: Removed some global namespace pollution.
- Updated the "GetUpgradedItemLevel" library. Now supports itemStrings with omitted zeros. Added more Timewarped Warforged IDs (thanks Cidrei).

Examiner Changes | 2015 Dec 07
------------------------------
- TOC version changed to 60200.
- Config: Check boxes should now work properly again.
- LibGearExam: Now detects the "Bonus Armor" stat.
- Updated the "GetUpgradedItemLevel" library. Should fix itemLevel issues with Timewarped items.

Examiner Changes | 2014 Oct 16
------------------------------
- TOC version has been changed to 60000 for the Warlords of Draenor 6.0.2 patch.
- Talents: Fixed for WoD. Modifed things around slightly to fit in a new tier of talents (tanks Netrox). Corrected the passed parameter to GetTalentInfo() as well as the expected returns. Now using the new/renamed constants, i.e. MAX_TALENT_TIERS.
- Guild: Fixed for WoD. Changed to work again with the removal of guild XP and level.
- Updated the "GetUpgradedItemLevel" library. This update will again properly find the correct itemLevel of upgraded items.
- Stats: Player defenses will now show the "PvP Power" stat. It was being tallied by LibGearExam before, but just not shown anywhere within Examiner.
- Localization: Updated the Korean file "koKR.lua". Thanks to fallendos.
- LibGearExam: The "PvP Power" string is now obtained from the translated global constant STAT_PVP_POWER.
- LibGearExam: Changed the itemLink patterns to be more robust and future-proof.

Examiner Changes | 2014 May 24
------------------------------
- Updated the "GetUpgradedItemLevel.lua" library, to include the new upgrade IDs.

Examiner Changes | 2014 Mar 01
------------------------------
- Removed some global namespace polution from the talents module.

Examiner Changes | 2014 Feb 24
------------------------------
- PvP: This module has been updated by ywfn, to work with all the PvP and battleground changes.

Examiner Changes | 2013 Sep 21
------------------------------
- Changed the TOC version to 50400 for the 5.4 patch.
- PvP: Updated this module to work with the now removed API functions. Arena data is no longer supported by WoW. Loading a cached unit that had Arena data stored, will still show however.
- GetUpgradedItemLevel: Patch 5.4: Added IDs 491 to 498 to the table.

Examiner Changes | 2013 May 22
------------------------------
- Changed the TOC version to 50300 for the 5.3 patch.
- Moved the GetUpgradedItemLevelFromItemLink() function into its own Lua file.
- GetUpgradedItemLevel: Patch 5.3: Added the 465/466/467 IDs (0/4/8 lvls) to the table.
- GetUpgradedItemLevel: Patch 5.2: Added the 470 ID (8 lvls) to the table.

Examiner Changes | 2013 Apr 19
------------------------------
- Changed the TOC version to 50200 for the 5.2 patch.
- LibGearExam: Removed a cause for a scan alert debug message. A string pattern match which had been marked obsolete, has again appeared in item tooltips from a new metagem.

Examiner Changes | 2013 Feb 19
------------------------------
- Changed the TOC version to 50100 for the 5.1 patch.
- Items should now show the proper item level when it has been upgraded.
- Removed the code that allowed the "Inspect" item of the drop down unit frame menu to be available at any distance. It is believed this was causing taint. If you want to keep inspecting at any range, assign a keybinding to Examiner.
- GearList: No longer thinks that head pieces are a viable target for enchants.

Examiner Changes | 2012 Oct 30
------------------------------
- Added new "Delayed Scan" module. It should fix some of issues where not all items were loaded after inspecting someone.
- Feats: Fixed an error that often occurred when an achievement tooltip was showing, with the option "Special Achievement Tooltip" enabled.
- Talents: This module is back! It's still not perfect, and it has no glyph details yet.

Modules/pvp.lua
local ex = Examiner;
local cfg;

-- Module
local mod = ex:CreateModule(PVP,PLAYER_V_PLAYER);
mod.help = "Honor Details";
mod:CreatePage(false);
mod:HasButton(true);
mod.canCache = true;

-- Variables
local labels = {};

-- Data Variables
local hd = {};
local ad = { {}, {}, {} };

-- Config
mod:AddOption({ var = "pvpbutton", default = false, label = "Display PvP Tab", tip = "If enabled Examiner will display the PvP tab." });


--------------------------------------------------------------------------------------------------------
--                                           Module Scripts                                           --
--------------------------------------------------------------------------------------------------------

-- OnInitialize
function mod:OnInitialize()
	cfg = ex.cfg;
end

-- OnConfigChanged
function mod:OnConfigChanged()
	if cfg.pvpbutton then
		mod:HasButton(true);
	else
		mod:HasButton(false);
	end
end

-- OnInspect
function mod:OnInspect(unit)
	mod.rankBar:Hide();
	ex:RequestHonorData();
end

-- OnHonorReady
function mod:OnHonorReady()
	self:LoadHonorNormal();
end

-- ShowTab
function mod:ShowTab()
	mod:HasButton(false);
end

-- OnCacheLoaded
function mod:OnCacheLoaded(entry,unit)
	if (entry.Honor) then
		self:HasData(true);
		-- Honor
		for name, value in next, entry.Honor do
			hd[name] = value;
		end
		self:UpdateHonor();
	end
end

-- OnCache
function mod:OnCache(entry)
	if (self:CanCache()) and (next(hd)) then
		entry.Honor = CopyTable(hd);
	end
end

-- OnClearInspect
function mod:OnClearInspect()
	self:HasData(nil);
	-- Header
	self.rankIcon:Hide();
	-- Clear Honor
	wipe(hd);
	for i = 1, 34 do
		labels[i]:SetText("");
	end
end

-- Round numbers
function round(number, decimals)
    return (("%%.%df"):format(decimals)):format(number)
end

--------------------------------------------------------------------------------------------------------
--                                             PvP Stuff                                              --
--------------------------------------------------------------------------------------------------------

-- Load Honor Normal
function mod:LoadHonorNormal()
	self:HasData(true);
		-- Get Honor values
		hd.todayHK, hd.todayDK, hd.yesterdayHK, hd.yesterdayHonor, hd.thisweekHK, hd.thisweekHonor, hd.lastweekHK, hd.lastweekHonor, hd.lastweekStanding, hd.lifetimeHK, hd.lifetimeDK, hd.lifetimeRank = GetInspectHonorData();
		-- Get Rank Progress
		hd.rankProgress = round((GetInspectPVPRankProgress() * 100),2)
		-- Get Highest Rank Name
		hd.lifetimeHighestRankName = GetPVPRankInfo(hd.lifetimeRank)
		if not hd.lifetimeHighestRankName then
			hd.lifetimeHighestRankName = COMBAT_TEXT_NONE
		end
		-- Get Current Rank Name
		if ex.unit == "target" then
			hd.rankName, hd.rankNumber = GetPVPRankInfo(UnitPVPRank("target"));
		elseif ex.unit == "mouseover" then
			hd.rankName, hd.rankNumber = GetPVPRankInfo(UnitPVPRank("mouseover"));
		end
		if not hd.rankName then
			hd.rankName = COMBAT_TEXT_NONE
			hd.rankNumber = 0
			hd.rankProgress = 0
		end	
	-- Update
	self:UpdateHonor();
end

-- Honor Update
function mod:UpdateHonor()
	-- Show Rank Bar Progress
	mod.rankBar:SetValue(hd.rankProgress)
	mod.rankBar.Text:SetText(hd.rankName .. " (" .. RANK .. " " .. hd.rankNumber .. ")")
	mod.rankBar.tip = ACHIEVEMENT_CATEGORY_PROGRESS .. ": " .. hd.rankProgress .. "%"
	mod.rankBar:Show();
	
	-- Show Icon with Current Rank
	if (hd.rankNumber ~= 0) then
		self.rankIcon.texture:SetTexture("Interface\\PvPRankBadges\\PvPRank"..format("%.2d",hd.rankNumber));
		self.rankIcon.texture:SetTexCoord(0,1,0,1);
		self.rankIcon.tip = format("%s (Rank %d)",hd.rankName,(hd.rankNumber));
		self.rankIcon:Show();
	end
	
	-- Show Kills/Honor
	labels[1]:SetText(HONOR_TODAY .. ":");
	labels[3]:SetText(HONORABLE_KILLS);
	labels[4]:SetText("|cff80ff80" .. hd.todayHK);
	labels[5]:SetText(DISHONORABLE_KILLS);
	labels[6]:SetText("|cffff6060" .. hd.todayDK);
	labels[7]:SetText(HONOR_YESTERDAY .. ":");
	labels[9]:SetText(HONORABLE_KILLS);
	labels[10]:SetText("|cff80ff80" .. hd.yesterdayHK);
	labels[11]:SetText(HONOR);
	labels[12]:SetText("|cffffff5D" .. hd.yesterdayHonor);
	labels[13]:SetText(ARENA_THIS_WEEK .. ":");
	labels[15]:SetText(HONORABLE_KILLS);
	labels[16]:SetText("|cff80ff80" .. hd.thisweekHK);
	labels[17]:SetText(HONOR);
	labels[18]:SetText("|cffffff5D" .. hd.thisweekHonor);
	labels[19]:SetText(HONOR_LASTWEEK .. ":");
	labels[21]:SetText(HONORABLE_KILLS);
	labels[22]:SetText("|cff80ff80" .. hd.lastweekHK);
	labels[23]:SetText(HONOR);
	labels[24]:SetText("|cffffff5D" .. hd.lastweekHonor);
	labels[25]:SetText(HONOR_STANDING);
	labels[26]:SetText("|cffffff5D" .. hd.lastweekStanding);
	labels[27]:SetText(HONOR_LIFETIME .. ":");
	labels[29]:SetText(HONORABLE_KILLS);
	labels[30]:SetText("|cff80ff80" .. hd.lifetimeHK);
	labels[31]:SetText(DISHONORABLE_KILLS);
	labels[32]:SetText("|cffff6060" .. hd.lifetimeDK);	
	labels[33]:SetText(HONOR_HIGHEST_RANK);	
	labels[34]:SetText("|cffffff5D" .. hd.lifetimeHighestRankName);	
end


--------------------------------------------------------------------------------------------------------
--                                           Widget Creation                                          --
--------------------------------------------------------------------------------------------------------

-- Rank Icon
mod.rankIcon = CreateFrame("Frame",nil,mod.page);
mod.rankIcon:SetPoint("TOPLEFT",12,-16);
mod.rankIcon:SetWidth(18);
mod.rankIcon:SetHeight(18);
mod.rankIcon:EnableMouse(1);
mod.rankIcon:SetFrameStrata("HIGH");
-- mod.rankIcon:SetScript("OnEnter",function(self) GameTooltip:SetOwner(self,"ANCHOR_BOTTOMRIGHT"); GameTooltip:SetText(self.tip) end)
-- mod.rankIcon:SetScript("OnLeave",ex.HideGTT);
mod.rankIcon.texture = mod.rankIcon:CreateTexture(nil,"ARTWORK");
mod.rankIcon.texture:SetAllPoints();

-- Rank Bar
local rankStart = 0
local rankEnd = 100
mod.rankBar = CreateFrame("StatusBar",nil,mod.page);
mod.rankBar:SetPoint("TOPLEFT",17,-15);
mod.rankBar:SetWidth(200);
mod.rankBar:SetHeight(20);
mod.rankBar:EnableMouse(1);
mod.rankBar:SetBackdrop({bgFile = [[Interface\PaperDollInfoFrame\UI-Character-Skills-Bar]]})
mod.rankBar:SetBackdropColor(0.2, 0.2, 0.5, 0.45)
mod.rankBar:SetStatusBarTexture([[Interface\PaperDollInfoFrame\UI-Character-Skills-Bar]])
mod.rankBar:SetStatusBarColor(0.4, 0.4, 0.7)
mod.rankBar:SetMinMaxValues(rankStart, rankEnd)
mod.rankBar:SetScript("OnEnter",function(self) GameTooltip:SetOwner(self,"ANCHOR_BOTTOMRIGHT"); GameTooltip:SetText(self.tip) end)
mod.rankBar:SetScript("OnLeave",ex.HideGTT);
mod.rankBar.Text = mod.rankBar:CreateFontString(nil,"ARTWORK") 
mod.rankBar.Text:SetFont(GameFontNormal:GetFont(), 13, "OUTLINE")
mod.rankBar.Text:SetPoint("CENTER",0,0)
mod.rankBar.Text:SetTextColor(1,1,0);
mod.rankBar.Border = CreateFrame("Frame", nil, mod.rankBar)
mod.rankBar.Border:SetPoint("TOPLEFT", mod.rankBar, "TOPLEFT", -2, 2)
mod.rankBar.Border:SetPoint("BOTTOMRIGHT", mod.rankBar, "BOTTOMRIGHT", 2, -2)
mod.rankBar.Border:SetBackdrop({ 
	edgeFile = "Interface/Tooltips/UI-Tooltip-Border", 
	tile = false, edgeSize = 10, 
	insets = { left = -1, right = -1, top = -1, bottom = -1}
})
mod.rankBar.Border:SetFrameLevel(mod.rankBar:GetFrameLevel() -1)
mod.rankBar.Border:SetBackdropBorderColor(0.6, 0.6, 0.8, 0.4)
	
-- Honor Labels
for i = 1, 34 do
	local l = mod.page:CreateFontString(nil,"ARTWORK","GameFontHighlightSmall");
	l:SetWidth(125);
	labels[i] = l;
	-- Configure table left and right column
	if ((i - 1) % 2 == 0) then
		if (i == 1) then
			l:SetPoint("BOTTOMLEFT", mod.rankBar, "BOTTOMLEFT", -6, -25);
			labels[i]:SetTextColor(0.5,0.75,1);
			labels[i]:SetFont(GameFontNormal:GetFont(),11);	
		else 
			-- Titles and data
			if (i == 7 or i == 13 or i == 19 or i == 27) then
				l:SetPoint("BOTTOMLEFT", labels[i - 2], "BOTTOMLEFT", 0, -19);
				labels[i]:SetTextColor(0.5,0.75,1);
				labels[i]:SetFont(GameFontNormal:GetFont(),11);	
			else
				l:SetPoint("BOTTOMLEFT", labels[i - 2], "BOTTOMLEFT", 0, -11.5);
				labels[i]:SetFont(GameFontNormal:GetFont(),10);
			end
		end
		l:SetJustifyH("LEFT");
	else
		l:SetPoint("LEFT", labels[i - 1], "RIGHT");
		l:SetWidth(90);
		l:SetJustifyH("RIGHT");
	end
end